### PR TITLE
[Android/tflite] Support gpu delegate @open sesame 11/06 10:52 

### LINKED ISSUE #2852: [Android/tflite] Support gpu delegates in Android
REF: #2700 

- [X] Build tflite with gpu delegate
- [X] Test with nnstreamer in Android device

# tflite delegates benchmark results

- Tested on S20 (qualcomm)
- I checked average invoke latency and total overhead from the log of tflite subplugin:

    ```log
    XXXXXXX E/GLib: Invoke() is finished: 35049
    XXXXXXX E/GLib: 100 invoke average 33740, total overhead 211
    ```

## Image classfication model

Used model: mobilenet_quant_v1_224.tflite

- input: (uint8)1x224x224x3 = 150,528 bytes
- output: (uint8)1x1001 = 1,001 bytes
- total **600** frames

|delegates|avg. invoke latency (us)|total overhead (us)|
|---|---|---|
|NONE|48,155|786|
|NNAPI|5,554|608|
|GPU <br> opencl backend |13,228|1145|
|GPU <br> opencl backend, set priority MIN_LATENCY|9024|1000|
|GPU <br> gles backend, set priority MIN_LATENCY|Runtime Error|Runtime Error|

## Object detection model

Used model: ssd_mobilenet_v2_coco.tflite

- input: (float32)1x300x300x3 (1,080,000 bytes)
- output: (float32)1x1917x1x4, (float32)1x1917x91 (total 728,460 bytes)
- total **100** frames

|delegates|avg. invoke latency (us)|total overhead (us)|
|---|---|---|
|NONE|130,989|104|
|NNAPI|42,761|112|
|GPU <br> opencl backend |47,929|288|
|GPU <br> opencl backend, set priority MIN_LATENCY|33,680|211|
|GPU <br> gles backend, set priority MIN_LATENCY|113,187|208|

### PR DESCRIPTION
**Pull request by anyj0527**
- Tflite gpu delegate requires sdk version higher than 28. Bump up the minimum sdk version from 24 to 28
- Support gpu delegate for tflite-2.3.0
- Resolves #2852 
- REF: https://github.com/nnstreamer/nnstreamer-android-resource/pull/13

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: api/android/README.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ We assume that you already have experienced Android application developments wit
  * Host PC:
    * OS: Ubuntu 16.04 / 18.04 x86_64 LTS
    * Android Studio: Ubuntu version
-   * Android SDK: Min version 24 (Nougat)
+   * Android SDK: Min version 28 (Pie)
```
Comment (jaeyun-jung) [inline api/android/README.md:10]: plz update the document in example repo later.

**Review by jaeyun-jung — APPROVED** [review]

**Comment by jaeyun-jung** [comment]
To reviewers, tf-lite 2.3 with gpu-delegate https://github.com/nnstreamer/nnstreamer-android-resource/pull/13

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 2]
```diff
@@ -46,6 +46,11 @@
 #include <tensorflow/contrib/lite/kernels/register.h>
 #endif
 
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2 || TFLITE_VERSION_MINOR >= 14)
```
Comment (again4you) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:49]: Q. Does tf-lite 1.14 support GPU delegate? 
If not, need to update the check condition.

At this stage, '__ANDROID__' definition is OK.
After your patch is merged, I will update the code to support both andorid and tizen.

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:49]: Thanks for pointing out! I've just checked with 1.15.2.
It does support gpu delegate, but its feature seems premature: support gl backend only, does not support quantized model, ... and the client code is quite different.

@jaeyun-jung and I decided support this feature only for 2.3.0, and I updated the code. Please check!

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 3]
```diff
@@ -305,14 +317,26 @@ TFLiteInterpreter::loadModel (int num_threads, accl_hw accelerator)
...
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2)
+    /** set gpu delegate when accelerator set to GPU */
+    TfLiteGpuDelegateOptionsV2 options = TfLiteGpuDelegateOptionsV2Default();
+    options.experimental_flags = TFLITE_GPU_EXPERIMENTAL_FLAGS_NONE;
+    options.experimental_flags |= TFLITE_GPU_EXPERIMENTAL_FLAGS_ENABLE_QUANT;
+
+    gpu_delegate.reset (TfLiteGpuDelegateV2Create (&options));
+    setDelegate (gpu_delegate.get ());
+#else
+    ml_logw ("GPU delegate is available only in Android with tflite v2.3.0");
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:339]: If gpu delegate is available after tflite 2.3.0, is it okay to check TFLITE_VERSION_MAJOR only?

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:339]: In https://github.com/nnstreamer/nnstreamer-android-resource/tree/master/external, we provide v2.3.0 only whose major version equals 2. So I thought the major version check would be sufficient.

Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:339]: Well.. then it requires code changes whenever libraries with different version are added. I recommend you to consider forward-compatibility.

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc]: You're right, I updated the log message with `v2.3.0 or higher`. Thanks!

**Review by dongju-chae — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 4]
```diff
@@ -46,6 +46,14 @@
 #include <tensorflow/contrib/lite/kernels/register.h>
 #endif
 
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2 || TFLITE_VERSION_MINOR >= 14)
+#include <tensorflow/lite/delegates/nnapi/nnapi_delegate.h>
+#endif
+
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2 && TFLITE_VERSION_MINOR >= 3)
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:53]: Does this mean that TFLite 3.0 will not support gpu delegate? :)

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:53]: Would more specific check: `(TFLITE_VERSION_MAJOR == 2 && TFLITE_VERSIONM_MINOR >=3)` be reasonable?

Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:53]: ```((TFLITE_VERSION_MAJOR == 2 && TFLITE_VERSIONM_MINOR >=3) || TFLITE_VERSION_MAJOR > 2)```

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc]: Updated, Please check!

**Review by myungjoo — APPROVED** [review]

**Pull request merged**